File: 91-decode-ways/decode-ways.cpp
class Solution {
public:

// int numDecodings(string s) {
//     int n=s.length();
//     int dp[n+1];
//     dp[0]=1;
//     dp[1]=(s[0]=='0'?0:1);
//     for(int i=2;i<=n;i++)
//     {
//         if(s[i-1]!='0')
//         {
//             if(s[i-2]!='0' && stoi(s.substr(i-2,2))<=26)
//             {
//                 dp[i]=dp[i-1]+dp[i-2];
//             }
//             else dp[i]=dp[i-1];
//         }
//         else 
//         {
//             if(stoi(s.substr(i-2,2))>0 && stoi(s.substr(i-2,2))<=26) dp[i]=dp[i-2];
//             else dp[i]=0;
//         }
//     }
//     return dp[n];
// }



/*


    int fun(int i,string s,vector<int>& dp)
    {
        if(i>=s.length()) return 1;
        if(dp[i]!=-1) return dp[i];
        int ans=0;
        int op1= s[i]-'0';
        int op2=0;
        if(i<s.length()-1) op2=op1*10+s[i+1] - '0';
        if(op1>0) ans+=fun(i+1,s,dp);
        if(op1>0 && op2>0 && op2<=26) ans+=fun(i+2,s,dp);
        dp[i]=ans;
        return dp[i];
    }

    int numDecodings(string s) {
        vector<int> dp(101,-1);
        return fun(0,s,dp);
    }
*/


// backtracking

    int backtrack(string s,int start_index,vector<int>& memo)
    {
        if(start_index==s.length()) return 1;
        if(memo[start_index]!=-1) return memo[start_index];
        int ans=0;
        int num1 = stoi(s.substr(start_index,1));
        if(num1>0 && num1<27) ans+=backtrack(s,start_index+1,memo);
        if(num1>0 && start_index+1<s.length())
        {
            int num2 = stoi(s.substr(start_index,2));
            if(num2<27) ans+=backtrack(s,start_index+2,memo);
        }
        memo[start_index]=ans;
        return ans;
    }

    int numDecodings(string s) {
        vector<int> memo(s.length(),-1);
        return backtrack(s,0,memo);
    }

};

File: 91-decode-ways/decode-ways.java
class Solution {

    public int numDecodings(String s) {
        /*
        CHOICES:
        -> SINGLE DIGIT : 
            -> 1 - 9 = 1 choice
            -> 0 no choice

            choice at index += choice at index - 1
        -> DOUBLE DIGIT (if possible):
            -> 10 - 26 = 1 choice
            -> else = no choice

            choice at index += choice at index - 2
        */

        int n = s.length();
        int[] dp = new int[n+1];
        //"" -> empty (1 way to decode)
        dp[0] = 1;
        if(s.charAt(0)=='0') return 0;
        dp[1] = 1;
        for(int i=2;i<=n;i++)
        {
            int singleDigit = Integer.valueOf(s.substring(i-1,i));
            if(singleDigit>=1) dp[i] += dp[i-1];
            int doubleDigit = Integer.valueOf(s.substring(i-2,i));
            if(doubleDigit>=10 && doubleDigit<=26) dp[i] += dp[i-2];
        }
        return dp[n];
    }
























/*
    public int backtrack(String s, int index,int[] memo) {
        if(memo[index]!=-1) return memo[index];
        if(index == s.length())
        {
            memo[index] = 1;
            return 1;
        }
        int poss = 0;
        String num = ""+s.charAt(index);
        if(Integer.parseInt(num)>0) poss += backtrack(s,index+1,memo);
        if(index<s.length()-1)
        {
            num += s.charAt(index+1);
            if(Integer.parseInt(num)>9 && Integer.parseInt(num)<=26)
            poss += backtrack(s,index+2,memo);
        }
        memo[index] = poss;
        return poss;
    }

    public int numDecodings(String s) {
        int[] memo = new int[s.length()+1];
        Arrays.fill(memo,-1);
        return backtrack(s,0,memo);
        //dynamic programming
        /*
        char[] ch = s.toCharArray();
        int n = ch.length;
        int[] dp = new int[n+1];
        dp[0] = 1;
        if(ch[0]!='0') dp[1] = 1;
        for(int i=2;i<=n;i++)
        {
            String num = "";
            num += ch[i-1];
            if(Integer.parseInt(num)>0) dp[i]+=dp[i-1];
            num = ch[i-2] + num;
            if(Integer.parseInt(num)>9 && Integer.parseInt(num)<=26) dp[i]+=dp[i-2];
        }
        return dp[n];
        
    }
*/
}